torch/nn/modules/lazy.py

```diff
@@ -0,0 +1,195 @@
+    def _buffers(self):
+        ...
+
+    @property
+    def _non_persistent_buffers_set(self):
+        ...
+
+    @property
+    def _load_hook(self):
+        ...
+
+    @property
+    def _initialize_hook(self):
+        ...
+
+
+class LazyModuleMixin:
+    r"""A mixin for modules that lazily initialize parameters.
+
+    Unlike regular modules, subclasses of this module can initialize
+    parameters outside of the constructor (``__init__``).
+    This allows you to, for example, initialize parameters in ``forward``
+    method to determine the shape of the weight based on the initial input.
+    Be sure to run "dummy" ``forward`` once to initialize all parameters in case
+    that you need to access the weights before the actual "real" ``forward`` pass.
+
+    >>> lazy_module = torch.nn.LazyLinear(10)
+    >>> lazy_module.forward(input)  # Dummy forward call to initialize the module parameters
+    >>> assert type(lazy_module) is torch.nn.Linear
+
+    Note that lazy modules cannot validate if the shape is correct during
+    deserialization.  Also note that the initial weights values may become different
+    from the original (non-lazy) module even if the random seed is manually
+    configured, as the order of initialization is different from the original
+    one; especially, ``module.cuda()`` may cause the initialization to run on
+    a GPU.
+
+    The default value of lazy parameters are ``UninitializedParameter()``.
+
+    .. note:: A `LazyModule`, or a module containing a `LazyModule` requires an explicit
+    dummy forward call in the following scenarios.
+
+    * Access an :class:`UnitializedParameter` data or properties
+    >>> lazy_module.weight.shape
+
+    * Use an Optimizer that accesses Parameters data in their constructor 
+      (Currently only :class:`torch.optim.AdaGrad`) other optimizers work without
+      explicitly initializing the module
+
+    >>> optim = torch.optim.AdaGrad(lazy_module.parameters())
+
+    * Use :class:`torch.nn.DataParallel` or :class:`torch.nn.DistributedDataParallel`
+    >>> module_dp = torch.nn.DataParallel(lazy_module)
+
+    * Module reparametrization using :func:`torch.nn.utils.weight_norm` or :func:`torch.nn.utils.spectral_norm`
+    >>> torch.nn.utils.weight_norm(lazy_module)
+
+    * When dealing with non-default module initialization
+    >>> torch.nn.init.uniform_(lazy_module.weight)
+    """
+
+    # modules inheriting from this will change their __class__ to the specified
+    # one after they are fully initialized
+    cls_to_become = None
+
+    def __init__(self: _LazyProtocol, *args, **kwargs):
+        # Mypy doesnt like this super call in a mixin
+        super().__init__(*args, **kwargs)  # type: ignore
+        self._load_hook = self._register_load_state_dict_pre_hook(self._lazy_load_hook)
+        self._initialize_hook = self.register_forward_pre_hook(self._infer_parameters)
+
+    def _save_to_state_dict(self: _LazyProtocol, destination, prefix, keep_vars):
```

Comment:
Things have changed a little since this feature was first proposed. What's the value of these modules having a state dict?

Reply:
As @yf225 suggested, there are modules that will remain as lazy ones because there are no actual non-lazy modes that they can be converted to after the initialization. These kinds of modules should support serialization as well.

Reply:
That's extremely interesting. Sorry @yf225 I missed your question about this. Could you elaborate on your thinking? I'm worried about the complexities of serializing lazy modules.

Reply:
I completely understand the issue with saving the state of a lazy module, a lazy module that has not been initialized has no actual state, but the state of a non-lazy module can be loaded turning it into a non-lazy module.

I think that forbidding serializing a LazyModule (by throwing an exception when accessining an `UninitializedParameter` is fine. In this case a module that inherits from the lazy mixin but has no UninitializedParameters could still work. 
However, ideally, when the module its materialized, it should remove the LazyMixin and hooks from its `__class__.__bases__` and become a regular nn.Module instance. @yf225 what are your thoughts on this?

@albanD do you agree with raising an error when the state dict is requested?

Reply:
Per our other discussion, let's not change this code now, but we may want to revisit serialization in the future.